Add missing texture and TGA validation checks for Mapped Images

DevGeniusCode opened the pull request:
**This PR introduces validation checks for missing textures and TGA files in mapped image configurations. It ensures that all referenced textures and TGA files are correctly accounted for and reports missing or duplicate items in the respective directories. The changes include:**

1. **Mapped Image Parsing:**
   - Scans `.ini` mapped files for texture and TGA references, ensuring that all referenced textures are validated and duplicates are reported.

2. **Missing TGA Files:**
   - Compares the list of TGA files from the `.ini` files against a provided CSV file and reports any missing TGA files that are not listed.

3. **Missing Textures in WND Files:**
   - Scans `.wnd` files for references to images and textures. Missing textures that are referenced in WND files but not present in the mapped images are reported.

4. **Output Files:**
   - Generates several output files, including:
     - `mapped_textures_list.txt`: A list of all texture-TGA mappings.
     - `duplicate_mapped_textures.txt`: A list of duplicate texture mappings.
     - `missing_tga_files.txt`: A list of missing TGA files.
     - `missing_wnd_textures.txt`: A list of missing textures referenced in WND files.

5. **Configurable Scanning Options:**
   - **Scan Original or Edited Files:** Allows users to select whether to scan the original or edited files, by setting the `is_original` flag. In the case of edited files (`is_original = False`), the function also checks the TGA files located in the `txtur_folder_path` directory and includes them in the validation check. If any TGA file is missing in the CSV and is not found in the `txtur_folder_path` directory, it is reported as missing.
   - **Include or Exclude `HandCreatedMappedImages.ini`:** Enables users to choose whether to include the `HandCreatedMappedImages.ini` file in the scan. This is controlled by skipping or including the file based on user preference. (This file is not included now in the output files.)

DevGeniusCode commented on Patch104pZH/Design/Scripts/textures/generated_original/duplicate_mapped_textures.txt:
```diff
@@ -0,0 +1,43 @@
+HandCreatedMappedImages.INI LoadPageHuge
```
Refers to duplicated MappedImage textures name, usually the same texture name in different `ini` files. See for example `SSRadarVanScan`, appearing twice in `SUUserInterface512.INI` file

xezon replied:
Ok can you find out which definitions take precedence? We should then remove (or comment out) all shadowed definitions.

DevGeniusCode replied:
Of course, right now I'm just creating the scripts to investigate all the files and their contents

DevGeniusCode commented:
Optimized the script and added a check for invalid textures in W3D files. 
Additionally, I added the `get_textures` function in `w3dfilemanager.py` to extract the textures from W3D files.

xezon commented on Patch104pZH/Design/Scripts/textures/get_mapped_textures_info.py:
```diff
@@ -0,0 +1,147 @@
+        print(f"Error: The file {file_path} was not found.")
+        return None
+
+def get_tga_and_textures(folder_path):
```
This is a confusing name because tga are textures.

xezon commented on Patch104pZH/Design/Scripts/textures/get_mapped_textures_info.py:
```diff
@@ -0,0 +1,147 @@
+def get_tga_and_textures(folder_path):
+    texture_and_file = []  # List of (tga, texture)
+    textures = set()  # Set of unique textures
+    tga_files = set()  # Set of unique tga files
```
Why are textures and tga's treated separately? Can't we just use one "textures" for all?

DevGeniusCode replied:
The textures and TGA files are treated separately because the INI files contain references to both the texture name (e.g., `PowerBarGreen`) and the actual TGA file (e.g., `SCGameUserInterface512_001.tga`).

Here is an example of how the data is structured in the INI files:

```
MappedImage PowerBarGreen
  Texture = SCGameUserInterface512_001.tga
  TextureWidth = 512
  TextureHeight = 512
  Coords = Left:497 Top:77 Right:499 Bottom:83
  Status = NONE
End
```

In this case, `PowerBarGreen` is the texture name, and `Texture` refers to the actual TGA file. I agree that the distinction might not be immediately clear, and I would be happy to consider any suggestions to make it more intuitive.

xezon replied:
Ok. I suggest rename "texture" to "mapped_image" and "tga_file" to "texture".

xezon commented on Patch104pZH/Design/Scripts/textures/get_mapped_textures_info.py:
```diff
@@ -0,0 +1,147 @@
+                        else:
+                            duplicate_textures.append((filename, texture))
+
+    return texture_and_file, list(textures), list(tga_files), duplicate_textures
```
This is quite complicated. Perhaps return a class instead of this tuple.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/missing_tga_files.txt:
Better: "missing_mapped_textures.txt"

DevGeniusCode replied:
I changed the file name to `missing_textures_files` because the name you suggested will be more appropriate for a later stage, when I print the mapped textures present in the INI files but not in the mapping files.
Maybe it would be better to call it `invalid_mapped_textures`? I have to admit that I’m a bit confused myself within the code between the actual TGA files and the mapped textures, as both are textures, and the variable names are getting mixed up.

DevGeniusCode replied:
I think we should use the word "textures" for the files (like TGA\DDS\W3D files) and "images" for the mapped textures (the ones from the mapping process).

xezon replied:
`invalid_mapped_textures` would be better then.

xezon commented on Patch104pZH/Design/Scripts/textures/get_mapped_textures_info.py:
Perhaps rename to "find_texture_errors.py"

DevGeniusCode commented:
optimized the code and updated various functions/variables names for better clarity.

DevGeniusCode commented:
1.  **Texture/Image Validation**:
    *   Implements a check for invalid textures and images. This identifies resources referenced in `.ini` files but not found in the file system, indicating broken or missing assets.
    *   Adds a check for unused textures and images. This finds resource files present in the file system but without corresponding references in any `.ini`/`.wnd`/`.w3d` file, suggesting potential dead assets.

2.  **Interactive Main Menu**:
    *   Enhances the main script with a user-interactive dialog. Users can now choose which function to execute: either the new validation checks or the existing texture and image extraction process.

DevGeniusCode commented:
I think I've now covered all the necessary validation for images and textures.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/invalid_images_in_ini_files.txt:
```diff
@@ -0,0 +1,11 @@
+4
```
I expect that this number 4 is difficult to find with a text search. Perhaps add a bit more context to this record to make it identifiable.

DevGeniusCode replied:
The regular expression is a bit lazy, I will improve it.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/invalid_textures_in_ini_files.txt:
```diff
@@ -0,0 +1,80 @@
+aiblocktile
```
Could you perhaps add a bit more context to the line, aka what it belongs to?

DevGeniusCode replied:
Can do this, but it will overload the file, as an invalid name can appear in multiple files. You can easily find them by searching in VSC or another tool.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_images_files.txt:
Better name would be "unused_mapped_images"

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_images_files.txt:
Why are all mapped images lower case? I suggest print them out in their original case.

DevGeniusCode replied:
There are too many differences due to capitalization, so I converted everything to lowercase to find only the ones that are truly invalid because no files with those names exist.
Will print with the original name

xezon replied:
Yes. Use lower case for comparison where applicable, but print original.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_images_files.txt:
```diff
@@ -0,0 +1,393 @@
+goodstatsicon
+helpbox-bottom
+helpbox-middle
+helpbox-top
```
I know that the helpbox textures are definitely used in game. They are referenced in game code.

DevGeniusCode replied:
Do you have any idea how to find them? Just need to check if these names exist in the dat/exe file

xezon replied:
The strings are certainly in the executable (game.dat). Not sure if plain ascii text search will suffice but you can try. Make sure the string ends on a 0 (null terminator).

xezon replied:
One potential problem here is capitalization of string. Either do a case insensitive search or get the search string in the correct capitalization. The mapped image name is "Helpbox-middle" and not "helpbox-middle".

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_images_files.txt:
```diff
@@ -0,0 +1,393 @@
+honorstreak_500
+honorstreak_b
+honorstreak_g
+honorstreak_s
```
I expect that these honor medals are referenced in game code. So a lot of mapped images here may actually be used.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_textures_files.txt:
```diff
@@ -0,0 +1,4648 @@
+cbngasst_tank_r
+cbngasst_tank_rs
+cbngasst_tank_s
+cbngasst_wall
```
I am confident to say that the majority of the textures listed here are referenced in w3d models. This needs rework.

DevGeniusCode replied:
This is after checking the textures in w3d as well, just to be safe I'll check again.

xezon replied:
Perhaps there is an error in the search method. I know that these textures are used because I have seen them in the w3d viewer on models.

Make sure to combine Generals 1.08 and Zero Hour 1.04 files because they both provide unique assets.

DevGeniusCode replied:
That's right, I didn't extract the w3d files for Generals 1.08. Will do.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_original/invalid_wnd_images.txt:
"invalid_mapped_images_in_wnd_files"

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/missing_textures_files.txt:
I think this file needs a better name.

xezon reviewed (commented):
I think the txt file names need an improvement pass to clarify what they refer to.

DevGeniusCode commented:
Improvements and fixes, added checking for unused textures/images in the game.dat file, and added the SCPatchUserInterface512.INI file that was missing from the repo. the reports were updated and also include Generals 1.08 files.

Also, saw your YouTube review of my code. Yes, I know regex isn't everyone's cup of tea, but without it I'd be drowning in manual parsing. It's a necessary tool for this job, Oh, and a little secret: I'm also using AI to craft those regex monsters, so really, we are on the same boat! 😉

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_mapped_images.txt:
```diff
@@ -0,0 +1,303 @@
+Rank_MajorElite
+Rank_Major_China
+Rank_Major_GLA
+Rank_Major_USA
```
I think it is very suspicious that so many of these Rank images are claimed unused.

My guess is that they are used, but they are assembled by combining words, and therefore not found as whole words in the executable file.

So for example there could be "Rank_Major" + "_USA", "Rank_Major" + "_GLA", etc...

DevGeniusCode replied:
I suggest leaving 'unused' as I have no idea how to deal with it, and focusing on invalid references.

xezon replied:
I suggest document clearly that the unused mapped images are not entirely trustable because of the caveat above. There probably are a few unused ones but they would need to be found by manually reviewing the list.

DevGeniusCode replied:
I looked into it a bit, and it seems to use Rank_[level]_[army]

But I'll add a warning at the beginning of the file, there are too many cases to handle in code.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_mapped_images.txt:
I expect that most textures listed here are actually used, but are not easily found because they are assembled of parts depending on conditions.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_textures.txt:
Why is this list so much longer than the other?

DevGeniusCode replied:
The edited list only scanned the edited folder, not the content of files that exist in the original folder such as w3d files extracted from big.

xezon replied:
The GameFilesEdited builds on top of the original. Can you perhaps program it in a way that it does not show false errors by considering the original underlying files?

DevGeniusCode commented on Patch104pZH/GameFilesOriginalZH/Data/INI/MappedImages/TextureSize_512/SCPatchUserInterface512.INI:
What are the 3 folders actually?

xezon replied:
It is described here:
https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Patch104pZH#readme

DevGeniusCode commented on Patch104pZH/Design/Scripts/textures/generated_original/unused_textures.txt:
```diff
@@ -0,0 +1,2298 @@
+NVSupplyTk_d3
+NVSupplyTk_d4
+NVSupplyTk_d5
+NVSupplyTk_d6
```
Indeed, they are w3d files that are in csv but I didn't find any references to them. Should I separate w3d files from tga/dds files?

xezon replied:
Yes I suggest create a unused_textures.txt and unused_w3d.txt

DevGeniusCode commented on Patch104pZH/Design/Scripts/textures/generated_original/invalid_textures_in_ini_files.txt:
```diff
@@ -0,0 +1,107 @@
+CBHigh01_SG
+CBHigh01_SN
+CBHigh01_SNG
+CBHigh02_DS
```
If I understood correctly, these are w3d file models, and if so, then w3d files with these names should exist, which they don't.

DevGeniusCode replied:
For example `CBHangr02_DSNG` with incorrect letter composition, which causes issue #2377

xezon replied:
I don't understand this statement.

So for example we have:
```
Object MogadishuShop01Bib
  ; *** ART Parameters ***
  Draw = W3DModelDraw ModuleTag_01
    ConditionState = NONE
      Model = CBModis03_TR
    End
    ConditionState = DAMAGED
      Model = CBModis03_TRD
    End
    ConditionState = REALLYDAMAGED
      Model = CBModis03_TRD
    End
    ConditionState = RUBBLE
      Model = CBModis03_TRD
    End
  End
```
All these are Models. I suggest put these into a "invalid_models_in_ini_files.txt"

DevGeniusCode commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_mapped_images.txt:
```diff
@@ -0,0 +1,303 @@
+SNUAttack
+SSChevron1S
+SSChevron2S
+SSChevron3S
```
I think what you see in the game is `SSChevron3L`, meaning the ones with the big L suffix and not the small S suffix.

xezon replied:
Ok that is a fair assumption.

DevGeniusCode commented:
Update:
Separation of textures and models, improved reports.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_original/invalid_textures_in_ini_files.txt:
```diff
@@ -1,86 +1,11 @@
-PSAirCarrier_NS
-PSCarRapt_D1B
 SCCScroll
 SMUDGE
```
I think this refers to some special surface in the executable. Perhaps search game.dat as well.

DevGeniusCode replied:
This appears in `Data\INI\ParticleSystem.ini` at ` ParticleName = SMUDGE.tga`,

xezon commented on Patch104pZH/Design/Scripts/textures/find_mappad_images_and_textures_errors.py:
```diff
@@ -0,0 +1,473 @@
+# These combined names might not appear as complete strings in the executable file,
+# causing them to be falsely flagged as unused.
+# Manual review of the list is required to verify the actual usage of some images.
+##########################################
```
Perhaps omit this line make it look cleaner.

xezon commented on Patch104pZH/Design/Scripts/textures/find_mappad_images_and_textures_errors.py:
```diff
@@ -0,0 +1,473 @@
+
+def write_to_file(output_path, data, format_func=None, warning=False):
+    warning_message = '''
+##########################################
```
Perhaps omit this line make it look cleaner.

xezon commented on Patch104pZH/Design/Scripts/textures/find_mappad_images_and_textures_errors.py:
```diff
@@ -0,0 +1,473 @@
+    textures, models = get_textures_models_from_csv(folder_paths['csv'])
+    textures_files = textures.union(
+        get_files_from_folder(folder_paths['textures'])).union(
+        get_files_from_folder(folder_paths['ani']))
```
I wonder if we can classify ani files as cursors instead of textures to show them in separate lists.

DevGeniusCode commented on Patch104pZH/Design/Scripts/textures/generated_edited/unused_textures.txt:
```diff
@@ -872,612 +84,19 @@ Manic Aggression
-ParaLandSK
 Red Rock
 Rocky Rampage
 Rogue Agent
```
come form `maps/Rogue Agent/Rogue Agent.tga`, will filter maps

DevGeniusCode commented:
Update: fix script name and maps references and remove `ani` files

xezon commented on Patch104pZH/Design/Scripts/textures/find_mapped_images_and_textures_errors.py:
```diff
@@ -0,0 +1,476 @@
+        'wnd': f'{edited_folder_paths}/Window',
+        'w3d': f'{edited_folder_paths}/Art/W3D',
+        'ani': f'{edited_folder_paths}/Data/Cursors',
+        'textures': f'{edited_folder_paths}/Art/Textures',
```
Does this script also consider the language specific assets at
```
Data/English/Art/Textures
Data/English/Art/W3D
Data/German/Art/Textures
Data/German/Art/W3D
...
```
This logic would become more intricate if this script also tests if specific files are missing in specific languages (which does happen in the original).

I think it would not be strictly necessary because we already fixed the language specific asset errors, but if it is no big hassle it could be added to this script.

DevGeniusCode replied:
Yes, I will update this, just to point out that I don't have the language files other than English to research them. I scraped all file paths with relevant extensions from the file, so maybe instead of adding them, it's better to filter out the language files? This way they won't appear in lists of unused files.

xezon replied:
Language packs are here:
https://www.gentool.net/download/languages/

xezon commented on Patch104pZH/Design/Scripts/textures/find_mapped_images_and_textures_errors.py:
```diff
@@ -0,0 +1,476 @@
+        'wnd': f'{edited_folder_paths}/Window',
+        'w3d': f'{edited_folder_paths}/Art/W3D',
+        'ani': f'{edited_folder_paths}/Data/Cursors',
+        'textures': f'{edited_folder_paths}/Art/Textures',
```
Note that we also have a GameFilesOptional folder that adds optional assets on top of the core assets (GameFilesEdited). Perhaps it would be a consideration to also add that, but that would add another layer of complexity.
https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Patch104pZH/GameFilesOptional

Additionally, we also have assets that are generated via Mod Builder and defined in the json files:
https://github.com/TheSuperHackers/GeneralsGamePatch/blob/main/Patch104pZH/ModBundleCoreItems.json
https://github.com/TheSuperHackers/GeneralsGamePatch/blob/main/Patch104pZH/ModBundleOptionalItems.json

For example:
```json
                            {
                                "source": "Art/Textures/GenerateMip/cbapt01_n.psd",
                                "target": "Art/Textures/cbapt01_sn.dds"
                            },
                            {
                                "source": "Art/Textures/GenerateMip/cbapt01_ng.psd",
                                "target": "Art/Textures/cbapt01_sng.dds"
                            },
                            {
                                "source": "Art/Textures/GenerateMip/cbapt01_dn.psd",
                                "target": "Art/Textures/cbapt01_dsn.dds"
                            },
                            {
                                "source": "Art/Textures/GenerateMip/cbapt01_dng.psd",
                                "target": "Art/Textures/cbapt01_dsng.dds"
                            },
```

DevGeniusCode replied:
I think the additional lists will only contribute to unused lists, while our main concern is fixing broken references that can disrupt the game engine.

xezon replied:
Yes ideally Optional files do not solely fix Core issues, but they could if we made mistakes (and I am sure we did).

DevGeniusCode commented:
- Adds `analyze_language_files` to scan language folders for unused/invalid assets.
- Generates individual text reports for each language.
- Assumes language-specific assets are not present in global folders.
[The language assets parameter is a dict containing models and textures that are only located inside the language folder.]

- Updated regex in Animation Models to correctly extract model names even when optional parameters are present.
[Results of research on relevant value locations are detailed in `Graphical_Assets_Location.md`. If there are any additional locations or patterns not currently documented in that file, please let me know, as it will help improve the accuracy of the reports.]

xezon commented on Patch104pZH/Design/Scripts/textures/Graphical_Assets_Location.md:
Perhaps use lowercase in file name like all other files in this folder.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/invalid_textures_in_ini_files.txt:
```diff
@@ -0,0 +1,22 @@
+EXRedSmokePuff
+EXSateliteHack1
+EXthunderRadar
+EXthunderWave
```
Why is this not written as "EXthunderWave.tga" as per the original value in the INI file?

DevGeniusCode replied:
Since the comparison was always without suffixes, so I did not capture them, I will include the suffixes for the reports.

xezon replied:
Would a "Texture.psd" also match for a "Texture.tga" ?

DevGeniusCode replied:
Indeed, assuming that `Mod Builder` converts it to dds/tga

xezon replied:
Typically yes, but Mod Builder can also build a "Texture.psd" as "AnotherTexture.dds", and I assume the script will not catch that.

Perhaps it would be worthwhile to check against files in `Patch104pZH\.Build\RawBundleItems` (the files generated before archiving into big files), but this would require building (all languages) before running the script and likely make it more complicated.

DevGeniusCode replied:
Will the `ModBundleOptionalItems.json` file be sufficient for processing?

xezon replied:
Currently `ModBundleCoreItems.json` and `ModBundleOptionalItems.json`.

You could try search through all files listed in `ModJsonFiles.json` and find texture and model targets that are not `*.ext`. Perhaps that would suffice.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/languages/French.txt:
```diff
@@ -0,0 +1,24 @@
+============== Unused Textures (INI) ==============
+
+French install
+SCSmShellUserInterface512
```
Why do these files have no file extension?

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/languages/German.txt:
```diff
@@ -0,0 +1,27 @@
+
+CItoxic
+CItoxic_B
+CItoxic_C
```
These 3 are textures, not models.

xezon replied:
It is a good question why textures are located in the w3d folder. I do not know what is up with that.

DevGeniusCode replied:
[The current update only includes `*.w3d` files.]

xezon replied:
These are definitely texture files.

DevGeniusCode replied:
I mean, do I need to scan them? Does the game engine scan them? Because right now my code only scans by folder paths.

xezon replied:
I do not know that. We would need to test if the game can load textures from that folder.

DevGeniusCode replied:
I took the file `Data\English\Art\Textures\Defeated.tga` and replaced it with the French version, it worked.
I moved it to the W3D folder, the English version appeared, meaning it didn't work.
I deleted it, now there is no `Data\English\Art\Textures\Defeated.tga` file at all, it still appeared in the English version, meaning the file is also in `game.dat`, but a hex scan yielded no results. I was confused.

xezon replied:
The texture is not in game.dat

Defeated.tga is still inside original big file that is underlying Patch files.

Reads as if texture is not loaded from W3D folder, which means these 3 files are misplaced in the German language.

DevGeniusCode replied:
I deleted all the big files after extracting.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_original/languages/Chinese.txt:
```diff
@@ -0,0 +1,22 @@
+
+None
+
+============== Invalid Models Assets ==============
```
"Invalid Models Assets" reads strange, because it has 2 plural words. Probably should be "Invalid Models"

xezon commented on Patch104pZH/Design/Scripts/textures/generated_original/languages/Chinese.txt:
```diff
@@ -0,0 +1,22 @@
+
+UITRST_SKNP
+
+============== Invalid Textures in Mapped Images Assets ==============
```
"Mapped Images Assets" reads strange, because it has 2 plural words. Probably should be "Mapped Images"

xezon commented on Patch104pZH/Design/Scripts/textures/generated_original/languages/Chinese.txt:
```diff
@@ -0,0 +1,22 @@
+
+None
+
+============== Invalid Models Assets ==============
```
"Invalid Models" is a confusing term for this, because UITRST_SKNP.w3d is missing in Chinese. It is not an invalid file, but simply missing. Could you explain why "Invalid" was chosen as the title?

DevGeniusCode replied:
The invalid list indicates references that exist in the ini files but no files exist for them. In the original folder, this is indeed an invalid reference, but in the edited folder, a file has been added and now the reference is valid (None).

DevGeniusCode replied:
what about `Invalid Languages Models references in INI files`

xezon replied:
How about "Unrecognized Models in INI files" ?

xezon replied:
Ok keep it then. I guess reader can infer the meaning eventually.

xezon commented on Patch104pZH/Design/Scripts/textures/generated_original/unused_textures.txt:
```diff
@@ -0,0 +1,930 @@
+PTXPALM03
+PTXPALM03._Btga
+PTXPALM03_B.tga
+PTYucca01
```
Why are these files listed without file extensions?

DevGeniusCode replied:
By the way, these two files also appear here without extensions, here is the full path

```
Art/Terrain/PTXPALM03._Btga.tga
Art/Terrain/PTXPALM03_B.tga.tga
```

xezon commented on Patch104pZH/Design/Scripts/textures/find_graphical_assets_errors.py:
```diff
@@ -0,0 +1,591 @@
+
+        elif function_choice == "8":
+            print("Calling function: analyze_language_files")
+            analyze_language_graphical_assets(folder_paths['ini'].replace('/Data/INI', ''), languages, language_assets, output_folder,
```
Graphical assets can be described as "art files", which arguably is simpler to understand because the game folder that contains textures and models is called "Art".

DevGeniusCode commented:
Added file extensions and improved naming.

Note that since we only catch with an extension, it was found that the `ParticleSystem.ini` file has an invalid parameter `ParticleName = GlassShard01` because it does not include a file extension (it is currently not included in the files because it is not included in the regular expression)

xezon commented on Patch104pZH/Design/Scripts/textures/generated_edited/languages/Brazilian.txt:
```diff
@@ -0,0 +1,22 @@
+
+None
+
+============== Invalid Languages Textures references in INI Mapped Images ==============
```
Languages Textures references are 3 plurals in a row. I think only the last word should be plural.

xezon commented on Patch104pZH/Design/Scripts/textures/find_art_files_errors.py:
```diff
@@ -0,0 +1,592 @@
+    return list(textures_in_w3d_files)
+
+
+def extract_art_files_from_ini_files(folder_path):
```
Perhaps move contents of this function into more smaller functions to decrease its size.

DevGeniusCode commented:
Improve `extract_art_files_from_ini_files` and spelling

TODO:
- add `ani` assets.
- add json blender assets, currently is assuming that Mod Builder converts it to dds/tga
